### PR TITLE
Allow saving generated datasets

### LINKED ISSUE #1478: Generated data sources do not work in new state files
* Sample Data -> Generate Constant Dataset
* Accept default parameters
* Save state
* Reset/Restart tomviz
* Load state... crash

### PR DESCRIPTION
**Pull request by mathturtle**
Fixes #1478.

### CODE REVIEW

File: tomviz/PythonGeneratedDatasetReaction.cxx [hunk 1]
```diff
@@ -541,31 +546,24 @@ void PythonGeneratedDatasetReaction::addDataset()
...
-  dataSource->setFileName(proxy->GetAnnotation(Attributes::LABEL));
-  LoadDataReaction::dataSourceAdded(dataSource, true /* default modules */,
-                                    false /* child */);
-}
-
-vtkSmartPointer<vtkSMSourceProxy>
-PythonGeneratedDatasetReaction::getSourceProxy(const QString& label,
-                                               const QString& script,
-                                               const int shape[3])
+DataSource* PythonGeneratedDatasetReaction::createGeneratedDataSource(
```
Comment (cryos) [inline tomviz/PythonGeneratedDatasetReaction.cxx]: I think shorten the member name to createDataSource, you already have the generated in the class name, and the name is already very long :-)

File: tomviz/PythonGeneratedDatasetReaction.h [hunk 2]
```diff
@@ -38,18 +38,15 @@ class PythonGeneratedDatasetReaction : public pqReaction
 
   void addDataset();
 
-  static vtkSmartPointer<vtkSMSourceProxy> getSourceProxy(const QString& label,
-                                                          const QString& script,
-                                                          const int shape[3]);
+  static DataSource* createGeneratedDataSource(
```
Comment (cryos) [inline tomviz/PythonGeneratedDatasetReaction.h]: As before - shorten to createDataSource - generated is already clear from the class name

File: tomviz/PythonGeneratedDatasetReaction.cxx [hunk 3]
```diff
@@ -129,33 +130,35 @@ class PythonGeneratedDataSource : public QObject
...
+
+    QJsonObject argsJson = QJsonObject::fromVariantMap(vmap);
+
+    QJsonArray size;
+    size.append(shape[0]);
+    size.append(shape[1]);
+    size.append(shape[2]);
+
+    QJsonObject description;
+    description["pythonScript"] = this->pythonScript;
```
Comment (cryos) [inline tomviz/PythonGeneratedDatasetReaction.cxx]: Let's just call it script, this was used in the python operator too, I would argue that the member variable should be renamed too, but fine leaving as-is

File: tomviz/PythonGeneratedDatasetReaction.cxx [hunk 4]
```diff
@@ -541,31 +546,24 @@ void PythonGeneratedDatasetReaction::addDataset()
...
-PythonGeneratedDatasetReaction::getSourceProxy(const QString& label,
-                                               const QString& script,
-                                               const int shape[3])
+DataSource* PythonGeneratedDatasetReaction::createGeneratedDataSource(
+  const QJsonObject& sourceInformation)
 {
-  PythonGeneratedDataSource generator(label);
-  generator.setScript(script);
+  PythonGeneratedDataSource generator(sourceInformation["label"].toString());
+  generator.setScript(sourceInformation["pythonScript"].toString());
```
Comment (cryos) [inline tomviz/PythonGeneratedDatasetReaction.cxx]: This should also be moved to script, I don't think we need to specify language, and if so would likely do so as another key-value pair.

Comment (mathturtle) [reply tomviz/PythonGeneratedDatasetReaction.cxx]: Makes sense

File: tomviz/DataSource.h [hunk 5]
```diff
@@ -70,7 +70,8 @@ class DataSource : public QObject
   /// Create a new dataSource not associated with a source proxy
   DataSource(const QString& label = QString(), DataSourceType dataType = Volume,
              QObject* parent = nullptr,
-             PersistenceState persistState = PersistenceState::Saved);
+             PersistenceState persistState = PersistenceState::Saved,
+             const QJsonObject* sourceInfo = nullptr);
```
[the remark this answers is not in the document]
Comment (mathturtle) [reply tomviz/DataSource.h]: I didn't realize there was a null-like state to the QJsonObject.  I wanted it to do nothing by default and so made it a pointer.  But you have a good point.  I'll change it to a const ref.

**Comment by mathturtle** [comment]
Updated and rebased on master now that #1480 is merged.

**Pull request merged**